File: python3/Stacks/next_largest_number_to_the_right.py
from typing import List


def next_largest_number_to_the_right(nums: List[int]) -> List[int]:
    res = [0] * len(nums)
    stack = []
    # Find the next largest number of each element, starting with the 
    # rightmost element.
    for i in range(len(nums) - 1, -1, -1):
        # Pop values from the top of the stack until the current 
        # value's next largest number is at the top.
        while stack and stack[-1] <= nums[i]:
            stack.pop()
        # Record the current value's next largest number, which is at 
        # the top of the stack. If the stack is empty, record -1.
        res[i] = stack[-1] if stack else -1
        stack.append(nums[i])
    return res

File: cpp/Stacks/next_largest_number_to_the_right.cpp
#include <vector>

std::vector<int> nextLargestNumberToTheRight(std::vector<int>& nums) {
    std::vector<int> res(nums.size(), 0);
    std::vector<int> stack;
    // Find the next largest number of each element, starting with the 
    // rightmost element.
    for (int i = nums.size() - 1; i >= 0; i--) {
        // Pop values from the top of the stack until the current
        // value's next largest number is at the top.
        while (!stack.empty() && stack.back() <= nums[i]) {
            stack.pop_back();
        }
        // Record the current value's next largest number, which is at
        // the top of the stack. If the stack is empty, record -1.
        res[i] = !stack.empty() ? stack.back() : -1;
        stack.push_back(nums[i]);
    }
    return res;
}